// forest-express/dist/services/apimap-sender.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var ApimapSender = /*#__PURE__*/function () {
  function ApimapSender(_ref) {
    var forestUrl = _ref.forestUrl,
        logger = _ref.logger,
        superagentRequest = _ref.superagentRequest;
    (0, _classCallCheck2["default"])(this, ApimapSender);
    this.forestUrl = forestUrl;
    this.superagentRequest = superagentRequest;
    this.logger = logger;
  }

  (0, _createClass2["default"])(ApimapSender, [{
    key: "handleResult",
    value: function handleResult(result) {
      if (!result) return;

      if ([200, 202, 204].includes(result.status)) {
        if (result.body && result.body.warning) {
          this.logger.warn(result.body.warning);
        }
      } else if (result.status === 0) {
        this.logger.warn('Cannot send the apimap to Forest. Are you online?');
      } else if (result.status === 404) {
        this.logger.error('Cannot find the project related to the envSecret you configured. Can you check on Forest that you copied it properly in the Forest initialization?');
      } else if (result.status === 503) {
        this.logger.warn('Forest is in maintenance for a few minutes. We are upgrading your experience in the forest. We just need a few more minutes to get it right.');
      } else {
        this.logger.error('An error occured with the apimap sent to Forest. Please contact support@forestadmin.com for further investigations.');
      }
    }
  }, {
    key: "send",
    value: function send(envSecret, apimap) {
      var _this = this;

      this.superagentRequest.post("".concat(this.forestUrl, "/forest/apimaps")).send(apimap).set('forest-secret-key', envSecret).end(function (_error, result) {
        _this.handleResult(result);
      });
    }
  }]);
  return ApimapSender;
}();

module.exports = ApimapSender;
