// forest-express/dist/integrations/stripe/index.js
"use strict";

var _ = require('lodash');

var logger = require('../../services/logger');

var Routes = require('./routes');

var Setup = require('./setup');

function Checker(opts, Implementation) {
  var integrationValid = false;

  function hasIntegration() {
    return opts.integrations && opts.integrations.stripe && opts.integrations.stripe.apiKey;
  }

  function isProperlyIntegrated() {
    return opts.integrations.stripe.apiKey && opts.integrations.stripe.stripe && opts.integrations.stripe.mapping;
  }

  function isIntegrationDeprecated() {
    var isIntegrationValid = opts.integrations.stripe.apiKey && opts.integrations.stripe.stripe && (opts.integrations.stripe.userCollection || opts.integrations.stripe.userCollection);

    if (isIntegrationValid) {
      logger.warn('Stripe integration attributes "userCollection" and "userField" are now deprecated, please use "mapping" attribute.');
      opts.integrations.stripe.mapping = "".concat(opts.integrations.stripe.userCollection, ".").concat(opts.integrations.stripe.userField);
    }

    return isIntegrationValid;
  }

  function isMappingValid() {
    var models = Implementation.getModels();
    var mappingValid = true;

    _.map(opts.integrations.stripe.mapping, function (mappingValue) {
      var collectionName = mappingValue.split('.')[0];

      if (!models[collectionName]) {
        mappingValid = false;
      }
    });

    if (!mappingValid) {
      logger.error("Cannot find some Stripe integration mapping values (".concat(opts.integrations.stripe.mapping, ") among the project models:\n").concat(_.keys(models).join(', ')));
    }

    return mappingValid;
  }

  function castToArray(value) {
    return _.isString(value) ? [value] : value;
  }

  function integrationCollectionMatch(integration, model) {
    if (!integrationValid) {
      return false;
    }

    var models = Implementation.getModels();

    var collectionModelNames = _.map(integration.mapping, function (mappingValue) {
      var collectionName = mappingValue.split('.')[0];

      if (models[collectionName]) {
        return Implementation.getModelName(models[collectionName]);
      }

      return null;
    });

    return collectionModelNames.indexOf(Implementation.getModelName(model)) > -1;
  }

  if (hasIntegration()) {
    if (isProperlyIntegrated() || isIntegrationDeprecated()) {
      opts.integrations.stripe.mapping = castToArray(opts.integrations.stripe.mapping);
      integrationValid = isMappingValid();
    } else {
      logger.error('Cannot setup properly your Stripe integration.');
    }
  }

  this.defineRoutes = function (app, model) {
    if (!integrationValid) {
      return;
    }

    if (integrationCollectionMatch(opts.integrations.stripe, model)) {
      new Routes(app, model, Implementation, opts).perform();
    }
  };

  this.defineCollections = function (collections) {
    if (!integrationValid) {
      return;
    }

    _.each(opts.integrations.stripe.mapping, function (collectionAndFieldName) {
      Setup.createCollections(Implementation, collections, collectionAndFieldName);
    });
  };

  this.defineFields = function (model, schema) {
    if (!integrationValid) {
      return;
    }

    if (integrationCollectionMatch(opts.integrations.stripe, model)) {
      Setup.createFields(Implementation, model, schema.fields);
    }
  };

  this.defineSerializationOption = function (model, schema, dest, field) {
    if (integrationValid && field.integration === 'stripe') {
      dest[field.field] = {
        ref: 'id',
        attributes: [],
        included: false,
        nullIfMissing: true,
        // TODO: This option in the JSONAPISerializer is weird.
        ignoreRelationshipData: true,
        relationshipLinks: {
          related: function related(dataSet) {
            return {
              href: "/forest/".concat(Implementation.getModelName(model), "/").concat(dataSet[schema.idField], "/").concat(field.field)
            };
          }
        }
      };
    }
  };
}

module.exports = Checker;
